### PR TITLE
Add read-only REST API endpoint for Datasets

### PR DESCRIPTION
**Pull request by ephraimbuddy**
A read-only endpoint for Datasets.

### CODE REVIEW

File: airflow/api_connexion/openapi/v1.yaml [hunk 1]
```diff
@@ -3373,6 +3418,49 @@ components:
...
           description: The permission resource
 
+    Dataset:
+      description: |
+        A dataset item.
+
+        *New in version 2.4.0*
+      type: object
+      properties:
+        dataset_id:
```
Comment (ashb) [inline airflow/api_connexion/openapi/v1.yaml]: Any reason to not call this `id`?

File: airflow/api_connexion/endpoints/dataset_endpoint.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+from airflow.api_connexion.types import APIResponse
+from airflow.security import permissions
+from airflow.utils.session import NEW_SESSION, provide_session
+
+
+@security.requires_access([(permissions.ACTION_CAN_READ, permissions.RESOURCE_DATASET)])
+@provide_session
+def get_dataset(dataset_id, session):
+    """Get a Dataset"""
+    dataset = session.query(Dataset).filter(Dataset.id == dataset_id).one_or_none()
```
Comment (ashb) [inline airflow/api_connexion/endpoints/dataset_endpoint.py]:
```suggestion
    dataset = session.query(Dataset).get(dataset_id)
```

File: airflow/api_connexion/openapi/v1.yaml (hunk 1 above)
Comment (ephraimbuddy) [reply airflow/api_connexion/openapi/v1.yaml]: Just to be uniform with our other models e.g connection_id

**Review by blag — APPROVED** [review]
LGTM.

Once we have a DatasetReference model, we should add links to the dags that depend on a dataset to the `/datasets/{dataset_id}` endpoint.

File: airflow/api_connexion/openapi/v1.yaml (hunk 1 above)
Comment (jedcunningham) [reply airflow/api_connexion/openapi/v1.yaml]: To be fair, `connection_id` is different than `Connection.id`, which is hidden from the user. Same with `DagRun.id` and `Variable.id`. This is the first internal id we are exposing to the user, I think.

Comment (jedcunningham) [reply airflow/api_connexion/openapi/v1.yaml]: Found 2 existing cases where surrogate id's are exposed, `import_error_id` and `event_log_id`. I've still pushed a commit to change it to `id` though.

**Review by dstandish — APPROVED** [review]

File: airflow/api_connexion/openapi/v1.yaml (hunk 1 above)
Comment (jedcunningham) [reply airflow/api_connexion/openapi/v1.yaml]: Merging this with `id`, we can always switch back to `dataset_id` later if we decide it should be that way instead.

**Review by jedcunningham — APPROVED** [review]

**Pull request merged**